### PR TITLE
feat(image): add image dataUri with base64

### LINKED ISSUE #1550: Add image dataUri with base64
### Clear and concise description of the problem

Currently `faker.image.dataUri` returns only a URL that contains a SVG

### Suggested solution

We would like to have an additional implementation that generates a URI containing base64 image

Example: `data:image/jpeg;base64,iVBORw0KGgoAAAANSUhEUgAAABkAAAAZCAYAAADE6YVjAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyJpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoV2luZG93cykiIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6MEVBMTczNDg3QzA5MTFFNjk3ODM5NjQyRjE2RjA3QTkiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6MEVBMTczNDk3QzA5MTFFNjk3ODM5NjQyRjE2RjA3QTkiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDowRUExNzM0NjdDMDkxMUU2OTc4Mzk2NDJGMTZGMDdBOSIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDowRUExNzM0NzdDMDkxMUU2OTc4Mzk2NDJGMTZGMDdBOSIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PjjUmssAAAGASURBVHjatJaxTsMwEIbpIzDA6FaMMPYJkDKzVYU+QFeEGPIKfYU8AETkCYI6wANkZQwIKRNDB1hA0Jrf0rk6WXZ8BvWkb4kv99vn89kDrfVexBSYgVNwDA7AN+jAK3gEd+AlGMGIBFDgFvzouK3JV/lihQTOwLtOtw9wIRG5pJn91Tbgqk9kSk7GViADrTD4HCyZ0NQnomi51sb0fUyCMQEbp2WpU67IjfNjwcYyoUDhjJVcZBjYBy40j4wXgaobWoe8Z6Y80CJBwFpunepIzt2AUgFjtXXshNXjVmMh+K+zzp/CMs0CqeuzrxSRpbOKfdCkiMTS1VBQ41uxMyQR2qbrXiiwYN3ACh1FDmsdK2Eu4J6Tlo31dYVtCY88h5ELZIJJ+IRMzBHfyJINrigNkt5VsRiub9nXICdsYyVd2NcVvA3ScE5t2rb5JuEeyZnAhmLt9NK63vX1O5Pe8XaPSuGq1uTrfUgMEp9EJ+CQvr+BJ/AAKvAcCiAR+bf9CjAAluzmdX4AEIIAAAAASUVORK5CYII=`

(Source: https://stackoverflow.com/a/41057998/6897682)

### Alternative

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ghost**
Adds a image type parameter `type` to `faker.image.dataUri`.

If `type` equals `'svg'`, it returns a URI that contains a svg image as before.
If `type` equals `'base64'`, it returns a URI that contains a base64 image as described in #1550.

Currently the base64 image is generated by converting a svg image. I am not sure, if that is an issue. As @matthewmayer pointed out in the issue, it would probably bloat the code if it should also return a base64 encoded jpg or png.

Additionally I added a few tests for the new parameter similar to the ones that already existed.

Closes #1550.

### CODE REVIEW

**Comment by ST-DDT** [comment]
There seems to be merge conflicts. Could you please fix them?

**Comment by ghost** [comment]
Sorry for the late reply. I fixed the merge conflicts.

**Comment by xDivisionByZerox** [comment]
**Team decision:**

We accept this feature. This can now be reviewed.

File: src/modules/image/index.ts [hunk 1]
```diff
@@ -348,15 +348,17 @@ export class ImageModule {
   }
 
   /**
-   * Generates a random data uri containing an svg image.
+   * Generates a random data uri containing an svg image or an base64 image.
```
Comment (matthewmayer) [inline src/modules/image/index.ts]: In what way is this a "random" data uri? It's a fixed output for a width:height:color?

Comment (ST-DDT) [reply src/modules/image/index.ts]: We could change the color to random? (No default)
(But IMO that's a different issue)

Would you like to create an issue for that?

File: src/modules/image/index.ts [hunk 2]
```diff
@@ -380,18 +382,29 @@ export class ImageModule {
        * @default 'grey'
        */
       color?: string;
+      /**
+       * The type of the image.
+       *
+       * @default 'svg'
+       */
+      type?: 'svg' | 'base64';
```
Comment (ST-DDT) [inline src/modules/image/index.ts]: Should we change the base64 to `svg-base64`?
That way we could add `png-base64` or other formats later on here as well.
Note: `png` without base64 is not possible, so IMO an encoding option wouldn't work here.

File: src/modules/image/index.ts [hunk 3]
```diff
@@ -380,18 +382,29 @@ export class ImageModule {
        * @default 'grey'
        */
       color?: string;
+      /**
+       * The type of the image.
+       *
+       * @default 'svg'
```
Comment (ST-DDT) [inline src/modules/image/index.ts]: Should we make this random in v9?

File: src/modules/image/index.ts (hunk 1 above)
Comment (ghost) [reply src/modules/image/index.ts]: Yes, I would like to create an issue for that and work on it.

Comment (matthewmayer) [reply src/modules/image/index.ts]: Let us discuss at https://github.com/faker-js/faker/issues/2307

Comment (ghost) [reply src/modules/image/index.ts]: Oh. You were faster than me.

I fixed/updated the JSDoc comment.

File: src/modules/image/index.ts (hunk 2 above)
Comment (ghost) [reply src/modules/image/index.ts]: I changed it. Shall I create an issue for this?

File: src/modules/image/index.ts [hunk 4]
```diff
@@ -348,15 +348,17 @@ export class ImageModule {
...
    *
    * @param options Options for generating a data uri.
    * @param options.width The width of the image. Defaults to `640`.
    * @param options.height The height of the image. Defaults to `480`.
    * @param options.color The color of the image. Defaults to `grey`.
+   * @param options.type The type of the image. Defaults to `svg`.
    *
    * @example
    * faker.image.dataUri() // 'data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http...'
+   * faker.image.dataUri({ type: 'base64' }) // 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3...'
```
Comment (ST-DDT) [inline src/modules/image/index.ts]:
```suggestion
   * faker.image.dataUri({ type: 'svg-base64' }) // 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3...'
```

**Review by ST-DDT — DISMISSED** [review]

File: src/modules/image/index.ts [hunk 5]
```diff
@@ -348,15 +348,17 @@ export class ImageModule {
   }
 
   /**
-   * Generates a random data uri containing an svg image.
+   * Generates a data uri containing an svg image or an base64 image.
```
Comment (matthewmayer) [inline src/modules/image/index.ts]:
```suggestion
   * Generates a data uri containing an svg image, optionally base64-encoded.
```

**Review by matthewmayer — DISMISSED** [review]

File: src/modules/image/index.ts (hunk 5 above)
Comment (ST-DDT) [reply src/modules/image/index.ts]: I'm not sure about this. 
If we change this method in v9 to return a random format, then I'm not sure whether I would consider this `optionally`.
Also if we add `png`s/other binary formats in the future they must be base64 encoded AFAIK (although this is more of a theoretical point at this time).

File: src/modules/image/index.ts (hunk 3 above)
Comment (ST-DDT) [reply src/modules/image/index.ts]: I created a separate/related issue for that:

- https://github.com/faker-js/faker/issues/2314

File: src/modules/image/index.ts (hunk 5 above)
Comment (matthewmayer) [reply src/modules/image/index.ts]: How about

`Generates a data uri containing an URL-encoded SVG image or a Base64-encoded SVG image.`

Comment (ST-DDT) [reply src/modules/image/index.ts]: Works for me.

**Comment by ST-DDT** [comment]
I just noticed, that `dataUri` doesn't have seeded tests, so I created a PR for that.

Blocked by:

- https://github.com/faker-js/faker/pull/2317

**Comment by ST-DDT** [comment]
We added some (forgotten) seeded tests to ensure the values generated by this method are reproducible.
Please extend them to include the type option.

https://github.com/faker-js/faker/blob/b468d5fb32b12cdb9cd4645e0bbe252825dc518a/test/modules/image.spec.ts#L71-L82

File: test/modules/image.spec.ts [hunk 6]
```diff
@@ -495,5 +495,43 @@ describe('image', () => {
       });
       expect(dataUri).toMatchSnapshot();
     });
+
```
Comment (Shinigami92) [inline test/modules/image.spec.ts:500]: line 77 should get updated, should it?

File: src/modules/image/index.ts [hunk 7]
```diff
@@ -380,18 +382,29 @@ export class ImageModule {
        * @default 'grey'
        */
       color?: string;
+      /**
+       * The type of the image.
+       *
+       * @default 'svg'
+       */
+      type?: 'svg' | 'svg-base64';
```
Comment (Shinigami92) [inline src/modules/image/index.ts]: What do you think about `encoding: 'raw' | 'base64'`?
(question goes to both: other maintainers as well as author of the PR)

Comment (ST-DDT) [reply src/modules/image/index.ts]: `png` or other future variants don't support `raw` encoding.

Comment (Shinigami92) [reply src/modules/image/index.ts]: Oh... you are slightly correct
I think I initially thought more about to have `data:image/svg+xml` and `data:image/png base64` and so on...

Comment (Shinigami92) [reply src/modules/image/index.ts]: `type: 'svg-uri' | 'svg-base64'`
and later we would have the possibility for `'png-base64'` :shrug:

Comment (ghost) [reply src/modules/image/index.ts]: I find `'svg-uri'` or `'svg-xml'` better than `'svg'` because they are more descriptive.

Comment (ST-DDT) [reply src/modules/image/index.ts]: IMO both variants are xml, so I prefer `svg-uri`.

Maybe we should extend the jsdocs a bit too like.
`The type of the image to return. Consisting of the file extension and the used encoding.`

Something like that.

File: src/modules/image/index.ts [hunk 8]
```diff
@@ -348,15 +348,17 @@ export class ImageModule {
...
 
   /**
-   * Generates a random data uri containing an svg image.
+   * Generates a data uri containing an URL-encoded SVG image or a Base64-encoded SVG image.
    *
    * @param options Options for generating a data uri.
    * @param options.width The width of the image. Defaults to `640`.
    * @param options.height The height of the image. Defaults to `480`.
    * @param options.color The color of the image. Defaults to `grey`.
+   * @param options.type The type of the image. Defaults to `svg`.
```
Comment (ST-DDT) [inline src/modules/image/index.ts]:
```suggestion
   * @param options.type The type of the image. Defaults to `'svg-uri'`.
```

File: test/modules/image.spec.ts [hunk 9]
```diff
@@ -495,5 +497,43 @@ describe('image', () => {
...
+    it('should return a blank svg data', () => {
```
Comment (ST-DDT) [inline test/modules/image.spec.ts]:
```suggestion
    it('should return a blank svg-uri data', () => {
```

File: test/modules/image.spec.ts [hunk 10]
```diff
@@ -495,5 +497,43 @@ describe('image', () => {
...
+      const dataUri = faker.image.dataUri({
+        width: 456,
+        height: 789,
+        type: 'svg-uri',
+      });
+      expect(dataUri).toMatchSnapshot();
+    });
+
+    it('should return a background color svg data URI', () => {
```
Comment (ST-DDT) [inline test/modules/image.spec.ts]:
```suggestion
    it('should return a background color svg-uri data URI', () => {
```

File: src/modules/image/index.ts [hunk 11]
```diff
@@ -348,15 +348,17 @@ export class ImageModule {
   }
 
   /**
-   * Generates a random data uri containing an svg image.
+   * Generates a data uri containing an URL-encoded SVG image or a Base64-encoded SVG image.
    *
    * @param options Options for generating a data uri.
    * @param options.width The width of the image. Defaults to `640`.
    * @param options.height The height of the image. Defaults to `480`.
    * @param options.color The color of the image. Defaults to `grey`.
```
Comment (ST-DDT) [inline src/modules/image/index.ts:356]: This is missing the `'`s as well, but that will be changed in #2316 anyway.

File: src/modules/image/index.ts [hunk 12]
```diff
@@ -348,15 +348,17 @@ export class ImageModule {
   }
 
   /**
-   * Generates a random data uri containing an svg image.
+   * Generates a data uri containing an URL-encoded SVG image or a Base64-encoded SVG image.
```
Comment (ST-DDT) [inline src/modules/image/index.ts]:
```suggestion
   * Generates a random data uri containing an URL-encoded SVG image or a Base64-encoded SVG image.
```

Although this is not random yet, that will be fixed as part of #2316.

**Review by ST-DDT — COMMENTED** [review]
Sorry, for all the back and forth.

File: src/modules/image/index.ts [hunk 13]
```diff
@@ -380,18 +382,35 @@ export class ImageModule {
...
 
-    const rawPrefix = 'data:image/svg+xml;charset=UTF-8,';
-    return rawPrefix + encodeURIComponent(svgString);
+    const rawPrefix =
+      type === 'svg-uri'
+        ? 'data:image/svg+xml;charset=UTF-8,'
+        : 'data:image/svg+xml;base64,';
+    return type === 'svg-uri'
+      ? rawPrefix + encodeURIComponent(svgString)
+      : rawPrefix + Buffer.from(svgString).toString('base64');
```
Comment (ST-DDT) [inline src/modules/image/index.ts]: 🤔 

```suggestion
    return type === 'svg-uri'
        ? `data:image/svg+xml;charset=UTF-8,${encodeURIComponent(svgString)}`
        : `data:image/svg+xml;base64,${Buffer.from(svgString).toString('base64')}`;
```

**Review by ST-DDT — APPROVED** [review]

**Review by Shinigami92 — APPROVED** [review]

**Pull request merged**